File: src/main/java/cpsc224/creatures/CreatureAI.java
package cpsc224.creatures;

import java.util.Random;
import java.util.List;
import java.util.ArrayList;

public class CreatureAI {
    
    private List<Double> weaponWeights;
    private double totalWeight;
    
    /**
     * Instantiates a creatureAI
     * @param creature the creature
     * @param weaponWeights the weapon weights
     */
    public CreatureAI(Creature creature){
        weaponWeights = new ArrayList<Double>();
        totalWeight = 0.0;
        for (int i = 0; i < creature.getInventory().getWeapons().size(); i++){
            weaponWeights.add(1.0);
            totalWeight+=1;
        }     
    }

    public int calculateMove() {
        Random rand = new Random();
        double num = rand.nextDouble() * totalWeight;
        int index = 0;
        for (Double weaponWeight: weaponWeights){
            num -= weaponWeight;
            if (num <= 0){
                return index;
            }
            index++;
        }
        return 0;
    }
}

File: web/src/game/creatures/CreatureAI.ts
import type { Creature } from './Creature';

export class CreatureAI {
  private creature: Creature;
  private weaponWeights: number[];
  private totalWeight: number;

  constructor(creature: Creature) {
    this.creature = creature;
    this.weaponWeights = [...creature.getWeaponWeights()];
    this.totalWeight = this.weaponWeights.reduce((a, b) => a + b, 0);
  }

  calculateMove(): number {
    const weapons = this.creature.getInventory().getWeapons();
    let total = this.totalWeight;
    const weights = [...this.weaponWeights];

    for (let i = 0; i < weapons.length && i < weights.length; i++) {
      if (weapons[i].getMove().getUses() === 0) {
        total -= weights[i];
        weights[i] = 0;
      }
    }

    if (total <= 0) return 0;

    let roll = Math.random() * total;
    for (let i = 0; i < weights.length; i++) {
      roll -= weights[i];
      if (roll <= 0) return i;
    }
    return 0;
  }
}
